Add Build Image for dotnetcore3.1

chrisoverzero opened the pull request:
...and Get `make pr` to Work

*Issue #, if available:* #13

*Description of changes:*

Adds a `dotnetcore3.1` build image in the style of the other build images – pulling its base from Public ECR as Python 3.9 does. The pre-invocation of `dotnet` is from LambCI, and attributed appropriately.

*Notes:*

- It took some effort to to get `make pr` working correctly. I tried to keep the changes for that to a minimum.
- `pytest` wants the unit tests to be run in random order. I don't know enough about it to know whether that plugin is loaded for everyone or just for me, so I did not add it.
- I don't know how these get mirrored to Public ECR, but it should end up there, too. Not only on Docker Hub.
- I set `AWS_EXECUTION_ENV=AWS_Lambda_dotnetcore3.1` for the build image, but that should be set in the emulation image, or whatever it is that's used for `sam local invoke`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

chrisoverzero committed:
Add Build Image for dotnetcore3.1

...and Get `make pr` to Work

chrisoverzero committed:
Correct CodeUri Path for `dotnetcore3.1`

chrisoverzero commented:
I noticed now that other pull requests go against the `develop` branch. But the contributing guidelines say this:

> You are working against the latest source on the _main_ branch.

If I should rebase and point this elsewhere, please do let me know.

hoffa commented on build-image-src/Dockerfile-dotnetcore31:
```diff
@@ -0,0 +1,70 @@
+RUN curl -L https://dot.net/v1/dotnet-install.sh | bash -s -- -c 3.1 -i "${DOTNET_ROOT}" && \
+  mkdir /tmp/warmup && \
+  cd /tmp/warmup && \
+  dotnet new && \
```
`dotnet new` by itself shows the command help; is that expected? A comment here on why this is needed would also be helpful.

hoffa reviewed (commented):
Thanks for submitting this PR!

> If I should rebase and point this elsewhere, please do let me know.

We're indeed developing against `develop`, sorry for the confusion! I've updated the docs.

chrisoverzero committed:
Address Pull Request Feedback

- Suppress First-Time Run Messages
- Redirect Nuget Warmup Output
- Justify Nuget Warmup Procedure

chrisoverzero commented:
Ecch, I'm not super adroit at rebasing across fork and branch, but I believe that's done it.

Thanks so much!

chrisoverzero committed:
Suppress First-Time Output Better

Somehow, I failed to notice that the very next line was also setting
environment variables.

sriram-mv commented on build-image-src/Dockerfile-dotnetcore31:
```diff
@@ -0,0 +1,72 @@
+    NUGET_XMLDOC_MODE=skip
+
+# Warm up the nuget cache once now for faster startup on each use.
+RUN curl -L https://dot.net/v1/dotnet-install.sh | bash -s -- -c 3.1 -i "${DOTNET_ROOT}" && \
```
Should https://www.nuget.org/packages/Amazon.Lambda.Tools also be installed within the build image?

aws-lambda-builders always and checks and installs the latest, but this will make it so that its usable outside of AWS SAM CLI

chrisoverzero replied:
I had that thought (and that’s how my team uses build images), but we’ve [encountered problems](https://github.com/aws/aws-lambda-builders/issues/126) running `sam build` when it wants to upgrade the installed version of Amazon.Lambda.Tools. I was the one who [added it to LambCI](https://github.com/lambci/docker-lambda/pull/187) in the first place, and I wanted not to repeat the same mistake here. 

I can try to see if it all works in this build image, for sure, though.

chrisoverzero replied:
Yes, this fails in the same way as before. I added the installation of an old version of Amazon.Lambda.Tools to the Dockerfile:

```sh
dotnet tool install --global Amazon.Lambda.Tools --version 5.0.0
```

…and tried to update it from the running container (in the same way as aws-lambda-builders does):

```sh
bash-4.2# dotnet tool install -g Amazon.Lambda.Tools --ignore-failed-sources
Tool 'amazon.lambda.tools' failed to update due to the following:
Failed to uninstall tool package 'amazon.lambda.tools': Invalid cross-device link
```

I did come up with a way to have it all work, though, and without major trickiness. Fresh push in a moment.

chrisoverzero committed:
Enable Installation of Amazon.Lambda.Tools

...Which Is Compatible with aws-lambda-builders

(a.k.a. `sam build`)

sriram-mv commented on build-image-src/Dockerfile-dotnetcore31:
```diff
@@ -0,0 +1,85 @@
+  cd / && \
+  rm -rf /tmp/warmup /tmp/NuGetScratch /tmp/.dotnet
+
+# Now we do something tricky. Installing Amazon.Lambda.Tools here as a --global tool will
```
Thanks for the comments! It adds a ton of context on why this is necessary!

hoffa commented on build-image-src/Dockerfile-dotnetcore31:
```diff
@@ -0,0 +1,85 @@
+#   write to the same layer. No cross-mount writes.
+# - Running `dotnet lambda` after installing or upgrading (either directly or via running
+#   `sam build`) will pick up the newly installed, globally installed version of the tool.
+ENV PATH=~/.dotnet/tools:$PATH
```
Why is this necessary? FWIW, `dotnet-lambda` is in `/var/lang/bin`, which is already in the `PATH` through the parent image.

chrisoverzero replied:
_Please do let me know if I can make the explanation in the file any clearer after reading this._

`dotnet tool` can't upgrade (nor uninstall, I guess) tools installed in earlier Docker layers. As I mentioned in the conversation with @sriram-mv earlier, if we install a version of "Amazon.Lambda.Tools" in the docker image:

```sh
dotnet tool install --global Amazon.Lambda.Tools --version 5.0.0
```

…and some tool comes along which unconditionally tries to install and/or update `Amazon.Lambda.Tools`, you get this:

```sh
bash-4.2# dotnet tool install -g Amazon.Lambda.Tools --ignore-failed-sources
Tool 'amazon.lambda.tools' failed to update due to the following:
Failed to uninstall tool package 'amazon.lambda.tools': Invalid cross-device link
```

(That tool is `aws-lambda-builders`.)

Adding the global tools installation directory to the path _ahead_ of where we installed it here with `--tools-path` allows direct use of `dotnet lambda` _and_ use of `sam build`. When `sam build` runs `dotnet tool install -g Amazon.Lambda.Tools --ignore-failed-sources`, that `-g` one doesn't try to overwrite the `--tool-path` one and it takes precedence since it's earlier on the path.

hoffa replied:
Gotcha, thanks for the thorough explanation!

hoffa commented on build-image-src/Dockerfile-dotnetcore31:
```diff
@@ -0,0 +1,85 @@
+#   `sam build`) will pick up the newly installed, globally installed version of the tool.
+ENV PATH=~/.dotnet/tools:$PATH
+
+RUN dotnet tool install --tool-path "${DOTNET_ROOT}" Amazon.Lambda.Tools --verbosity quiet
```
nit: the output is typically squashed to a single line, so probably no need to lower verbosity. Also allows using `docker build --progress plain` to increase verbosity without touching files:

```suggestion
RUN dotnet tool install --tool-path "${DOTNET_ROOT}" Amazon.Lambda.Tools
```

chrisoverzero replied:
Oh, I did this because of the comment earlier about the `dotnet new` output. I'll happily remove this.

hoffa replied:
I see! For the `dotnet new` comment I was curious on whether it was supposed to do something, as the output didn't suggest it (but it seems it initializes a cache as you pointed out). If it's purely about output, then keeping it at whatever the default is seems fine.

chrisoverzero committed:
Remove Pointless Verbosity Reduction

Co-authored-by: Chris Rehn <crehn@outlook.com>

hoffa commented:
The image is now available as [`public.ecr.aws/sam/build-dotnetcore3.1`](https://gallery.ecr.aws/sam/build-dotnetcore3.1). 🎉  Thanks a lot for your contribution!
